swarm/storage/resource.go

```diff
@@ -0,0 +1,529 @@
+	}
+
+	// get our blockheight at this time and the next block of the update period
+	currentblock, err := self.getBlock()
+	if err != nil {
+		return nil, err
+	}
+	nextblock := getNextBlock(rsrc.startblock, currentblock, rsrc.frequency)
+
+	// start from the last possible block period, and iterate previous ones until we find a match
+	// if we hit startblock we're out of options
+	version := uint64(1)
+	for nextblock > rsrc.startblock {
+		key := self.resourceHash(rsrc.ensname, nextblock, version)
+		chunk, err := self.Get(key)
+		if err == nil {
+			// check if we have versions > 1. If a version fails, the previous version is used and returned.
+			log.Trace("rsrc update version 1 found, checking for version updates", "nextblock", nextblock, "key", key)
+			for {
+				newversion := version + 1
+				key := self.resourceHash(rsrc.ensname, nextblock, newversion)
+				newchunk, err := self.Get(key)
+				if err != nil {
+					// rsrc update data chunks are total hacks
+					// and have no size prefix :D
+					err := self.verifyContent(chunk.SData)
+					if err != nil {
+						return nil, err
+					}
+					// update our rsrcs entry map
+					rsrc.lastblock = nextblock
+					rsrc.version = version
+					rsrc.data = make([]byte, len(chunk.SData)-SIGNATURE_LENGTH)
+					rsrc.updated = time.Now()
+					copy(rsrc.data, chunk.SData[SIGNATURE_LENGTH:])
+					log.Debug("Resource synced", "name", rsrc.name, "key", chunk.Key, "block", nextblock, "version", version)
+					self.resourceLock.Lock()
+					self.resources[name] = rsrc
+					self.resourceLock.Unlock()
+					return rsrc, nil
+				}
+				log.Trace("version update found, checking next", "version", version, "block", nextblock, "key", key)
+				chunk = newchunk
+				version = newversion
+			}
+		}
+		log.Trace("rsrc update not found, checking previous period", "block", nextblock, "key", key)
+		nextblock -= rsrc.frequency
+	}
+	return nil, fmt.Errorf("no updates found")
+}
+
+// Adds an actual data update
+//
+// Uses the data currently loaded in the resources map entry.
+// It is the caller's responsibility to make sure that this data is not stale.
+//
+// A resource update cannot span chunks, and thus has max length 4096
+func (self *ResourceHandler) Update(name string, data []byte) (Key, error) {
+
+	// can be only one chunk long minus 65 byte signature
+	if len(data) > MAX_CHUNK_DATA {
+		return nil, fmt.Errorf("Data overflow: %d / %d bytes", len(data), 4096-SIGNATURE_LENGTH)
+	}
+
+	// get the cached information
+	self.resourceLock.Lock()
+	defer self.resourceLock.Unlock()
+	resource, ok := self.resources[name]
+	if !ok {
+		return nil, fmt.Errorf("No such resource")
+	} else if resource.updated.IsZero() {
+		return nil, fmt.Errorf("Invalid resource")
+	}
+
+	// get our blockheight at this time and the next block of the update period
+	currentblock, err := self.getBlock()
+	if err != nil {
+		return nil, err
+	}
+	nextblock := getNextBlock(resource.startblock, currentblock, resource.frequency)
```

Comment:
Small tip, feel free to ignore: a `nextBlock := self.getNextBlock()` line would be simpler instead of line 371-376 and line 298-303.
You would just need a getNextBlock() function similar getBlock() on ResourceHandler, the current getBlock function is not used in any other context anyway.

Reply:
getblock is used in `SetResource` :/